pkg/operator/migration_aws_status/controller.go

```diff
@@ -0,0 +1,131 @@
+	configv1listers "github.com/openshift/client-go/config/listers/config/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	operatorv1helpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	clusterConfigNamespace = "kube-system"
+	clusterConfigName      = "cluster-config-v1"
+	clusterConfigKey       = "install-config.yaml"
+)
+
+// MigrationAWSStatusController migrates the existing `infrastructure.config.openshift.io/v1` `cluster` objects
+// for AWS to include the `AWSPlatformStatus` based on the InstallConfig stored in the ConfigMap `kube-system/cluster-config-v1`.
+// BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1814332
+// The controller reads the configmap for the `install-config.yaml` and then creates a `AWSPlatformStatus` and updates the infrastructure object with these values.
+// Here are the values that are migrated by the controller:
+// - `.status.platformStatus.type` = `.status.platformType`
+// - `.status.platformStatus.aws.region` = AWS region from InstallConfig's `.platform.aws.region`
+//
+// The controller does no-op for non-AWS platforms.
+// It uses the `.status.platformType` from infrastructure object to identify the platform.
+type MigrationAWSStatusController struct {
+	infraClient     configv1client.InfrastructureInterface
+	infraLister     configv1listers.InfrastructureLister
+	configMapClient corev1client.ConfigMapsGetter
+}
+
+// NewController returns a MigrationAWSStatusController
+func NewController(operatorClient operatorv1helpers.OperatorClient,
+	infraClient configv1client.InfrastructuresGetter, infraLister configv1listers.InfrastructureLister, infraInformer cache.SharedIndexInformer,
+	configMapClient corev1client.ConfigMapsGetter,
+	kubeSystemInformer cache.SharedIndexInformer,
+	recorder events.Recorder) factory.Controller {
+	c := &MigrationAWSStatusController{
+		infraClient: infraClient.Infrastructures(),
+		infraLister: infraLister,
+	}
+	return factory.New().
+		WithInformers(
+			operatorClient.Informer(),
+			infraInformer,
+			kubeSystemInformer,
+		).
+		WithSync(c.sync).
+		WithSyncDegradedOnError(operatorClient).
+		ResyncEvery(time.Minute).
+		ToController("MigrationAWSStatusController", recorder)
+}
+
+func (c MigrationAWSStatusController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	obji, err := c.infraLister.Get("cluster")
+	if errors.IsNotFound(err) {
+		syncCtx.Recorder().Warningf("MigrationAWSStatusController", "Required infrastructures.%s/cluster not found", configv1.GroupName)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	currentInfra := obji.DeepCopy()
+
+	if currentInfra.Status.Platform != configv1.AWSPlatformType || // not aws
```

Comment:
[`Status.Platform` is deprecated][1].  I'd rather have something like:

```go
var platformType configv1.PlatformType
if currentInfra.Status.PlatformStatus != nil && len(currentInfra.Status.PlatformStatus.Type) >  0 {
  platformType = currentInfra.Status.PlatformStatus.Type
} else {
  platformType = currentInfra.Status.Platform
}

if platformType != configv1.AWSPlatformType || // not AWS
    ... {
  ...
}
```

or some such.

[1]: https://github.com/openshift/api/blob/0f159fee64dbf711d40dac3fa2ec8b563a2aaca8/config/v1/types_infrastructure.go#L54-L57

Reply:
deprecated fields don't mean they will not be set... to keep backward compatibility that field will always be set. the clients should start moving to the new fields.

And since this is the migrator that is moving clusters from deprecated to new state, and therefore not using the .status.platformType feels weird.. i don't think your if-else guard makes sense here.

Reply:
+100 here - deprecated means "you should start using something new" not "the old one changes or goes away".

We DO NOT remove fields ever. :)

Reply:
I see trevor's comment more as pointing out that if the authoritative field is .status.platformstatus.type (omg, why pointers??), we should trigger on that field and stomp the deprecated one to whatever is correct.

I'd like to have a goal of transitioning this field away from a pointer.  We can do that by requiring it in every cluster in 4.5.  This code highlights what's wrong with pointers.  Dear lord.

Reply:
I understand @wking comment and i have already added similar conditonal to other controllers like kube_cloud_config, aws_service_endpoints.

I was trying to say that, for this specific controller the source of truth needs to be the deprecated field because it is trying to migrate the clusters away so that use of deprecated can be gone later on.

after this migration is applied to 4.5, we can make the move from pointer. so 4.6

Reply:
> I was trying to say that, for this specific controller the source of truth needs to be the deprecated field...

I would be very surprised if existing 4.5 consumers are checking the deprecated field first and then falling back to the new field if the deprecated field is unset.

Reply:
registry operator [only falls back to install-config if `PlatformStatus` is `nil`][1].

[1]: https://github.com/openshift/cluster-image-registry-operator/blob/24d4772d517571edccbd219c43f34954435c0c3b/pkg/storage/util/util.go#L76-L85

Reply:
@wking the registry operator is going the right thing. as it is supposed to used the new field and only when unset use the deprecated one.

but doesn't mean this migration needs to do that.

Reply:
> I would be very surprised if existing 4.5 consumers are checking the deprecated field first and then falling back to the new field if the deprecated field is unset.

again echoing the point from https://github.com/openshift/cluster-config-operator/pull/127#discussion_r412305448 https://github.com/openshift/cluster-config-operator/pull/127#discussion_r411780867

this deprecated field cannot go away unless we bump api version , and will always continue to be set for the life of the api. that how k8s backward compatibilty works.

Reply:
> but doesn't mean this migration needs to do that.

If this PR prefers the deprecated property while the registry prefers the new property, then in the (unlikely) event of divergence, this migration is going to move the registry to a new region (not actually sure what the registry operator does if the region changes under its feet).  I think we want to consistently have folks prefer the new property and fall back to the deprecated property, and we also want this logic to update the deprecated property if it diverges, so old tools that only consume the deprecated property don't see stale data.